src/credentials/providers/credentials.ts

```diff
@@ -113,7 +113,7 @@ export interface CredentialsProvider {
      * first use (in particular, credentials that may prompt, such as SSO/MFA,
      * should _not_ attempt to auto-connect).
      */
-    canAutoConnect(): boolean
+    canAutoConnect(): boolean | Promise<boolean>
```

Comment:
hmm.. does this mean the caller has to add `await` depending on the implementation?

Reply:
No it's ok to `await` a non-promise. Only downside is that tests/logic can be written towards specific instances of the interface which may have the narrowed synchronous type. And if everything is inferred and/or non-strict equality is used then you can introduce bugs if instance type is widened.

I'm gonna change it to only be `Promise<boolean>` as it's too easy to break things when doing this with `boolean`.